tensorflow/core/kernels/matrix_triangular_solve_op.cc

```diff
@@ -24,15 +20,117 @@ limitations under the License.
 #include "tensorflow/core/platform/logging.h"
 #include "tensorflow/core/platform/macros.h"
 #include "tensorflow/core/platform/types.h"
+#include "third_party/eigen3/Eigen/Core"
+
+#if GOOGLE_CUDA
+#include "tensorflow/core/platform/stream_executor.h"
+#endif // GOOGLE_CUDA
 
 namespace tensorflow {
 
+#if GOOGLE_CUDA
+namespace {
+template <typename Scalar>
+perftools::gputools::DeviceMemory<Scalar> AsDeviceMemory(
+    const Scalar* cuda_memory) {
+  perftools::gputools::DeviceMemoryBase wrapped(
+      const_cast<Scalar*>(cuda_memory));
+  perftools::gputools::DeviceMemory<Scalar> typed(wrapped);
+  return typed;
+}
+}  // namespace
+#endif  // GOOGLE_CUDA
+
 template <class Scalar>
 class MatrixTriangularSolveOp : public LinearAlgebraOp<Scalar> {
+public:
+    typedef LinearAlgebraOp<Scalar> Base;
+
+    explicit MatrixTriangularSolveOp(OpKernelConstruction* context)
+        : Base(context)
+        , lower_(true)
+        , adjoint_(false)
+    {
+        OP_REQUIRES_OK(context, context->GetAttr("lower", &lower_));
+        OP_REQUIRES_OK(context, context->GetAttr("adjoint", &adjoint_));
+    }
+
+    using TensorShapes = typename Base::TensorShapes;
+    using Matrix = typename Base::Matrix;
+    using MatrixMap = typename Base::MatrixMap;
+    using MatrixMaps = typename Base::MatrixMaps;
+    using ConstMatrixMap = typename Base::ConstMatrixMap;
+    using ConstMatrixMaps = typename Base::ConstMatrixMaps;
+
+    virtual void ValidateInputMatrixShapes(
+        OpKernelContext* context,
+        const TensorShapes& input_matrix_shapes) const final
+    {
+        Base::ValidateSquareSolver(context, input_matrix_shapes);
+    }
+
+    TensorShapes GetOutputMatrixShapes(
+        const TensorShapes& input_matrix_shapes) const final
+    {
+        return TensorShapes({ TensorShape({ input_matrix_shapes[0].dim_size(1),
+            input_matrix_shapes[1].dim_size(1) }) });
+    }
+
+    int64 GetCostPerUnit(const TensorShapes& input_matrix_shapes) const final
+    {
+        double rows = static_cast<double>(input_matrix_shapes[0].dim_size(0));
+        double num_rhss = static_cast<double>(input_matrix_shapes[1].dim_size(1));
+        double cost = rows * rows * num_rhss;
```

Comment:
This is the original tensorflow op code (see above). Possibly because double can represent larger numbers than int? I doubt anyone will try a matrix multiply large enough to overflow a 64 bit int though. Happy to change to int64.